dungeon/src/contrib/utils/DynamicCompiler.java

```diff
@@ -0,0 +1,165 @@
+ *       the package and class declaration inside the target source file. For example, if the class
+ *       is declared as {@code package my.riddle; public class Answer { ... }}, then the input must
+ *       be {@code "my.riddle.Answer"}.
+ *   <li><strong>Isolation via Custom ClassLoader:</strong> A custom class loader is used to load
+ *       freshly compiled classes from disk instead of reusing classes already loaded by the JVM.
+ *       This is essential for scenarios where the same class name may be compiled multiple times
+ *       with different contents (e.g., repeated puzzle attempts).
+ *   <li><strong>Classes must not be part of the core project:</strong> If the class to be loaded
+ *       dynamically is also part of the main project (i.e., already compiled and present on the
+ *       default classpath), the system class loader will load it first, and our custom loader will
+ *       be unable to override it. Therefore:
+ *       <ul>
+ *         <li>The class may exist as a source file in the project, but must not be directly
+ *             referenced or compiled with the project. <br>
+ *             <em>Rule of thumb:</em> The project must still compile and run without modification
+ *             if you delete the target file.
+ *         <li><strong>Relying on Abstractions:</strong> The main application should depend on
+ *             abstractions (such as interfaces, abstract classes, or {@code Supplier<?>}) to
+ *             support injecting implementations that are loaded dynamically at runtime. <br>
+ *             <br>
+ *             <em>Example:</em> Imagine a game where item effects can be coded by the user. The
+ *             {@code Item} class should not hardcode any effect logic. Instead, it could hold a
+ *             reference to a {@code Supplier<ItemEffect>} or implement an {@code ItemEffect}
+ *             interface. The Java source file submitted by the user would then implement this
+ *             interface or provide a valid {@code get()} method if using a supplier. The part of
+ *             the code that triggers the dynamic loading (e.g., a riddle evaluator or configuration
+ *             loader) would use {@code DynamicCompiler.compileAndLoad} to load the class and call a
+ *             setter on the item (e.g., {@code item.setEffect(...)}) to inject the behavior
+ *             dynamically. This pattern allows the main application to remain stable, while
+ *             enabling highly customizable runtime extensions.
+ *       </ul>
+ * </ul>
+ */
+public class DynamicCompiler {
+
+  /** Where to save temporary generated files. */
+  private static final String OUTPUT_PATH = "build/relection";
+
+  /**
+   * Compiles and loads a Java class from the specified source file.
+   *
+   * @param sourcePath Path to the source file containing the class.
+   * @param className Fully qualified class name (must match declaration inside file).
```

Comment:
brauche ich wirklich beides? den pfad einerseits, und andererseits nochmal den class-name? aus dem bauch heraus sollte das doch auf die selbe information herauslaufen bzw. das eine kann aus dem anderen abgeleitet werden, oder übersehe ich grad was?

Reply:
Sowas wie `MySpeedPotionClass.class.getName().replace(".", "/") + ".java";` sollte den String generieren. Oder geht das nur, wenn die Klasse bereits geladen ist?

Reply:
Ich brauche den Namen um die Klasse zu generieren und auf das Ding zuzugreifen (vermute ich). 
Man könnte sich das bestimmt auf den eingelesenen Quellcode ziehen (via regex) aber das ist mir irgendwie zu risky.... Bauchgefühl sagt das könnte es einige Edgecases geben

Reply:
Im Beispiel PR hattest du doch:
```java
  private static final IPath SOURCE_FILE = new SimpleIPath("src/hotload/MySpeedEffect.java");
  private static final String CLASS_NAME = "hotload.MySpeedEffect";
```
verwendet. 
Könnten wir nicht einfach sagen die Klasse ist der Path zur Source File? oder anders herum?

Reply:
> Könnten wir nicht einfach sagen die Klasse ist der Path zur Source File? oder anders herum?

Dann stellt sich aber die Frage an, ab wo beginnt das Vollständige Klassenname? Ich brauch ja auch das package. Alles nach `src`? Was ist wenn meine Datei nicht in `src` liegt? Was ist wenn das package zwar im Target-File angegeben ist, aber sich nicht im Pfad wiederspiegelt.

Ich find das ist alles ziemlich viel Engineering um einen relativ simplen Parameter zu vermeiden. idk

Reply:
nur der vollständigkeit halber: in gradle hast du den source-ordner (`sourceSets.main.java.srcDirs`). mit einer weiteren systemproperty könntest du das dann im java-code auslesen (`sourceDir = sourceSets.main.java.srcDirs.asPath` auf gradle-ebene)

Reply:
#1963